### ISSUE TITLE
recommendations for a jumpy loss curve on the validation data set

### ISSUE THREAD

**Issue by taylorreiter** [greenelab/generic-expression-patterns#116]
I just finished training my first VAE with sophie! I used 861 RNA seq samples of Pseudomonas aeruginosa strain PAO1 as my input compendium. 

(side note, it only took about an hour, and this computer only has 16 gb of ram 🎉)

Comparing the plot of my loss curve to [one you have this in this repo from pseudomonas array data](https://github.com/greenelab/generic-expression-patterns/blob/4e0855d94b3bc426868e1ebcc84516ca861f0dff/pseudomonas_analysis/logs/NN_2500_30/tybalt_2layer_30latent_hist.svg), I'm guessing my validation loss curve isn't supposed to jump around like this. 

![tybalt_2layer_300latent_hist](https://user-images.githubusercontent.com/23057646/161322007-9f2c7915-886a-45a0-b77f-d75882cc439e.svg)

My relevant config params were:
```
latent_dim      300
num_simulated   25
NN_architecture "NN_2500_30"
learning_rate   0.001
batch_size      10
kappa   .01
intermediate_dim        2500
epsilon_std     1.0
epochs  100
validation_frac .25
```
My questions are:
1. are these loss curves ok because they stabilize toward the later epochs?
2. if the validation loss curve is not ok, what are your recommendations for how to adjust the NN params to improve the loss curve?

**Comment by ajlee21** [comment]
Yay for getting SOPHIE running:) Now we just need to get it to train!

In general I usually tend to see more of an elbow looking curve like you linked to, so a few thoughts come to mind:
1. Did you check to make sure you normalized your input data?
2. Given the amount of variability you see in the validation loss I'm wondering if you need to reduce the `validation_frac` so that you have more to train on.
3. Since you see that the validation loss is lower than the training loss (i.e the model is underfit, you might need to train the model for longer)

**Comment by taylorreiter** [comment]
>     1. Did you check to make sure you normalized your input data?
I followed the code in this notebook [1_process_and_train_vae.ipynb](https://github.com/greenelab/generic-expression-patterns/blob/master/new_model_experiment/1_process_and_train_vae.ipynb), which uses this code:
```


process.normalize_compendium(
    processed_compendium_filename,
    normalized_compendium_filename,
    scaler_filename,
)
```

Is that sufficient normalization, or do I need to do something else before?

>     2. Given the amount of variability you see in the validation loss I'm wondering if you need to reduce the `validation_frac` so that you have more to train on.

Ok! I'll try this.

>     3. Since you see that the validation loss is lower than the training loss (i.e the model is underfit, you might need to train the model for longer)
ok! I can try this as well

**Comment by ajlee21** [comment]
Using the normalization code should be sufficient. I was just thinking if you wanted to check the output dataframe and look to see if the values looked as expected if you use something like `displot` on a few of the genes/columns

**Comment by taylorreiter** [comment]
ok I increased the epochs to 200, and dropped the `validation_frac` to 0.15.
![tybalt_2layer_300latent_hist](https://user-images.githubusercontent.com/23057646/161603420-cbdcfb94-19b0-4454-8c0f-54dff146dc1a.svg)

**Comment by ajlee21** [comment]
I think its heading in the right direction. Here are some additional thoughts I have, but perhaps time to loop in other Greene lab eyes:

1. Compared to array data, RNA-seq has more variance. Currently, normalization SOPHIE uses right now is to 0-1 scaled based on the min-max. If outliers are present (something I might expect to be the case for RNA-seq and something you'd be able to see if you plot the distribution of the data), then that means that the scaling is based on outliers and the biological signal/structure in the resulting 0-1 scaled data will be muted. This will make learning the signal challenging and so the model might learn an oversimplified representation of the data and the training loss might converge very fast. Its hard to tell if this is what is happening with the training loss here and doesn't explain the shape/very different magnitude of the validation loss.

As a future note: we could think about using a normalization like z-scaling instead for RNA-seq data. Though when we use min-max scaling for recount2 (human RNA-seq data), we don't see this same issue with the validation loss >> training loss so this is probably not the major issue in training. https://github.com/greenelab/generic-expression-patterns/blob/master/human_general_analysis/logs/NN_2500_30/tybalt_2layer_30latent_hist.svg

3.  What I more suspect is that this could be an issue with the learning rate (i.e. the size of the steps the model is taking). If the step size if too small or too larger, this could produce a lot of jumping as the model gets "stuck" in sub optimal regions. You could try increasing learning rate to 0.01 and decreasing it to 0.0001.

**Comment by taylorreiter** [comment]
Thanks for all of the feedback Alex! I tried a lot more training parameter combinations, and I think I finally got an ok loss curve. I wanted to document the training process/parameters/feedback that I got in this issue. 

My first set of loss curves, compared to alex's loss curves with Pa array data, are below
<img width="1026" alt="Screen Shot 2022-04-05 at 2 56 37 PM" src="https://user-images.githubusercontent.com/23057646/162533270-e8bb5457-fcc1-4c6a-8e5c-c1af4af011bd.png">

And here is a random plot of gene counts for one gene:
![image](https://user-images.githubusercontent.com/23057646/162533562-3116e755-ad91-44ba-8302-0a74405564fb.png)

after seeing these curves, @danich1 gave the following advice:
> Would recommend increasing the batch size. Seems like 10 allows too many iterations during training otherwise [@Alex Lee](https://greenelab.slack.com/team/U034L2SND32) suggestion for decreasing the learning rate makes sense.

When I asked for more clarification on this, @danich1 said:

> So iirc on my knowledge of gradient descent: when you are training a neural network a common problem is figuring how large your batch size should be. If the batch is too high then your network won’t make a whole lot weight changes (back propagation). On the other end if your batch size is too small your network will update weights a lot more often which could cause erratic val loss.

@ben-heil added to this by stating:
>  think David's hypothesis is the most likely cause. If that direction doesn't help, here are a few more hypotheses:
Since the training loss is smooth and the val loss is spiky, the swings may be driven by the data not the training.
Having a small val set may cause individual samples to have error spikes during training that aren't washed out by the other samples (which isn't necessarily a problem as long as the loss is decreasing and you're using early stopping or something)
Alternatively the issue could be out-of-distribution samples in the val set which is more worrisome
Also I think there's a way to split out the reconstruction and KL loss in Alex's VAE implementation in case that helps debug

Given this feedback, I tried many more param combinations
<img width="1295" alt="Screen Shot 2022-04-08 at 3 02 32 PM" src="https://user-images.githubusercontent.com/23057646/162537884-1a5bee5f-3c21-4d27-be3a-b68f00af2c89.png">

 
I thought the last curve was starting to look better, but was given the feedback that it's a bit odd to have the validation loss be lower than the training loss. I found this great twitter thread on this topic:

https://twitter.com/aureliengeron/status/1110839223878184960?s=20&t=W8nt0deKnwcOQv_Nfkj3cA

One of the points in the twitter thread is that the validation loss curve can be lower than training when there isn't enough validation data. Given this, I increased the validation fraction to .25 and I think ended up with a pretty good loss curve!
<img width="578" alt="Screen Shot 2022-04-07 at 6 09 38 AM" src="https://user-images.githubusercontent.com/23057646/162537783-4de6643f-bac0-4d2d-98a7-b1700a70597c.png">

**Comment by ajlee21** [comment]
Referenced this issue as a resource for users using SOPHIE in https://github.com/greenelab/sophie/pull/3
